internal/proxy/multi_app_conn.go

```diff
@@ -147,23 +152,38 @@ func (app *multiAppConn) killTMOnClientError() {
 		}
 	}
 
-	select {
-	case <-app.consensusConnClient.Quit():
-		if err := app.consensusConnClient.Error(); err != nil {
-			killFn(connConsensus, err, app.Logger)
-		}
-	case <-app.mempoolConnClient.Quit():
-		if err := app.mempoolConnClient.Error(); err != nil {
-			killFn(connMempool, err, app.Logger)
-		}
-	case <-app.queryConnClient.Quit():
-		if err := app.queryConnClient.Error(); err != nil {
-			killFn(connQuery, err, app.Logger)
-		}
-	case <-app.snapshotConnClient.Quit():
-		if err := app.snapshotConnClient.Error(); err != nil {
-			killFn(connSnapshot, err, app.Logger)
-		}
+	type op struct {
+		connClient stoppableClient
+		name       string
+	}
+
+	for _, client := range []op{
+		{
+			connClient: app.consensusConnClient,
+			name:       connConsensus,
+		},
+		{
+			connClient: app.mempoolConnClient,
+			name:       connMempool,
+		},
+		{
+			connClient: app.queryConnClient,
+			name:       connQuery,
+		},
+		{
+			connClient: app.snapshotConnClient,
+			name:       connSnapshot,
+		},
+	} {
+		go func(name string, client stoppableClient) {
+			client.Wait()
+			if ctx.Err() != nil {
+				return
+			}
+			if err := client.Error(); err != nil {
+				killFn(name, err, app.Logger)
+			}
```

Comment:
What happens if the context cancels? This won't cause client.Wait() to return right?

Reply:
if the context that was passed to the client when things started is canceled, then Wait will eventually return. 

The check of the context error on line 180 should cause us to not SIGTERM ourselves in this case (i.e. normal shut down)